File: 2025/day8.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../coords3d'

coords = $stdin.each_line.map do |line|
  Coords3D.from_a(line.chomp.split(',').map(&:to_i))
end.to_a

# @param coords [Array<Coords3D>]
# @return [Array(Integer, Integer)] Array of pairs of indexes in `coords`.
def edges_sorted_by_distance(coords:)
  edges = []

  coords.each_with_index do |a, i|
    ((i + 1)...coords.length).each do |j|
      b = coords[j]
      edges << [a.squared_distance(b), i, j]
    end
  end

  edges.sort_by!(&:first)
  edges.map! { |_, i, j| [i, j] }
end

# @param root [Integer]
# @param neighbours [Hash{Integer => Set<Integer>}]
# @return [Set<Integer>]
def connected_component(root:, neighbours:)
  frontier = Set.new([root])
  visited = Set.new

  until frontier.empty?
    visited.merge(frontier)
    next_frontier = Set.new
    frontier.each do |node|
      next_frontier.merge(neighbours[node])
    end
    next_frontier.subtract(visited)
    frontier = next_frontier
  end

  visited
end

link_count = coords.count < 30 ? 10 : 1000

edges = edges_sorted_by_distance(coords: coords)

shortest_edges = edges.take(link_count)
neighbours = Hash.new { Set.new }
shortest_edges.each do |a, b|
  neighbours[a] += [b]
  neighbours[b] += [a]
end

# Find connected components

remaining_nodes = Set.new((0...coords.count).to_a)
components = []

until remaining_nodes.empty?
  component = connected_component(root: remaining_nodes.first, neighbours: neighbours)
  remaining_nodes.subtract(component)
  components << component
end

puts components.map(&:count).max(3).inject(1, &:*)

# Part 2: Add more edges one by one until the entire graph is one component

component_of_node = {}
component_by_id = {}

components.each_with_index do |component, index|
  component_by_id[index] = component
  component.each { |node| component_of_node[node] = index }
end

remaining_edges = edges.drop(link_count).reverse

while (a, b = remaining_edges.pop)
  ac = component_of_node[a]
  bc = component_of_node[b]
  next if ac == bc

  # The edge merged two components, renumber the one with fewer nodes

  a_nodes = component_by_id[ac]
  b_nodes = component_by_id[bc]

  if a_nodes.count < b_nodes.count
    a_nodes.each { |node| component_of_node[node] = bc }
    component_by_id[bc].merge(component_by_id.delete(ac))
  else
    b_nodes.each { |node| component_of_node[node] = ac }
    component_by_id[ac].merge(component_by_id.delete(bc))
  end

  if component_by_id.count == 1
    # Only one component remains, done
    puts coords[a].x * coords[b].x
    break
  end
end

File: 2025/day8.swift
import Foundation

// MARK: -

struct Node: Hashable, Sendable {
    typealias Value = Int64

    let id: Int
    let x: Value
    let y: Value
    let z: Value

    init(id: Int, coordinates: [Value]) {
        self.id = id
        self.x = coordinates[0]
        self.y = coordinates[1]
        self.z = coordinates[2]
    }

    func squaredDistance(to other: Node) -> Value {
        let dx = x - other.x
        let dy = y - other.y
        let dz = z - other.z
        return dx * dx + dy * dy + dz * dz
    }
}

// MARK: - Algorithms

func edgesSortedByDistance(between nodes: [Node]) -> [(Int, Int)] {
    var result = [(distance: Int64, a: Int, b: Int)]()
    result.reserveCapacity(nodes.count * (nodes.count - 1) / 2)

    for a in nodes {
        for b in nodes.suffix(from: a.id + 1) {
            let distance = a.squaredDistance(to: b)
            result.append((distance: distance, a: a.id, b: b.id))
        }
    }

    result.sort { $0.distance < $1.distance }
    return result.map { ($0.a, $0.b) }
}

func connectedComponent(root: Int, neighbours: [Int: Set<Int>]) -> Set<Int> {
    var frontier: Set<Int> = [root]
    var visited = Set<Int>()

    while !frontier.isEmpty {
        visited.formUnion(frontier)
        var nextFrontier = Set<Int>()

        for node in frontier {
            if let neigh = neighbours[node] {
                nextFrontier.formUnion(neigh)
            }
        }

        nextFrontier.subtract(visited)
        frontier = nextFrontier
    }

    return visited
}

// MARK: - Input

var nodes: [Node] = []
while let line = readLine(strippingNewline: true) {
    let fields = line.split(separator: ",").compactMap { Node.Value($0) }
    if fields.count != 3 { fatalError("Invalid input: \(line)") }
    nodes.append(Node(id: nodes.count, coordinates: fields))
}

guard nodes.count > 0 else { fatalError("No input") }


// MARK: - Part 1

let linkCount = nodes.count < 30 ? 10 : 1000
let edges = edgesSortedByDistance(between: nodes)

let initialEdges = Array(edges.prefix(linkCount))
var neighbours: [Int: Set<Int>] = [:]
for (a, b) in initialEdges {
    neighbours[a, default: []].insert(b)
    neighbours[b, default: []].insert(a)
}

// Find all connected components

var remainingNodes = Set(nodes.map(\.id))
var components: [Set<Int>] = []

while let root = remainingNodes.first {
    let comp = connectedComponent(root: root, neighbours: neighbours)
    remainingNodes.subtract(comp)
    components.append(comp)
}

let sizes = components.map { $0.count }.sorted(by: >)
print(sizes.prefix(3).reduce(into: 1) { $0 *= $1 })

// MARK: - Part 2

var componentOfNode: [Int: Int] = [:]
var componentById: [Int: Set<Int>] = [:]

for (componentId, component) in components.enumerated() {
    componentById[componentId] = component
    for node in component {
        componentOfNode[node] = componentId
    }
}

let remainingEdges = edges.suffix(from: linkCount)

for (a, b) in remainingEdges {
    guard let componentA = componentOfNode[a], let componentB = componentOfNode[b] else {
        fatalError("No component for \(a) - \(b) edge")
    }
    if componentA == componentB { continue }

    // Two components got merged, move the smaller into the larger

    let nodesA = componentById[componentA]!
    let nodesB = componentById[componentB]!

    if (nodesA.count, componentA) < (nodesB.count, componentB) {
        for node in nodesA {
            componentOfNode[node] = componentB
        }
        componentById[componentB] = nodesB.union(nodesA)
        componentById.removeValue(forKey: componentA)
    } else {
        for node in nodesB {
            componentOfNode[node] = componentA
        }
        componentById[componentA] = nodesA.union(nodesB)
        componentById.removeValue(forKey: componentB)
    }

    if componentById.count == 1 {
        // The entire graph is now one component
        print(nodes[a].x * nodes[b].x)
        break
    }
}
